File: obligation-cordapp/java-source/src/test/java/net/corda/examples/obligation/contract/ObligationContractSettleTests.java
package net.corda.examples.obligation.contract;

import com.google.common.collect.ImmutableList;
import net.corda.core.contracts.Amount;
import net.corda.core.contracts.CommandAndState;
import net.corda.core.contracts.PartyAndReference;
import net.corda.core.identity.AbstractParty;
import net.corda.core.identity.CordaX500Name;
import net.corda.examples.obligation.Obligation;
import net.corda.examples.obligation.ObligationContract;
import net.corda.finance.contracts.asset.Cash;
import net.corda.testing.core.TestIdentity;
import org.junit.Test;

import java.util.Currency;

import static net.corda.examples.obligation.ObligationContract.OBLIGATION_CONTRACT_ID;
import static net.corda.finance.Currencies.*;
import static net.corda.testing.node.NodeTestUtils.ledger;

public class ObligationContractSettleTests extends ObligationContractUnitTests {
    private TestIdentity issuer = new TestIdentity(new CordaX500Name("MegaBank", "", "US"));
    private Byte defaultRef = Byte.MAX_VALUE;
    private PartyAndReference defaultIssuer = issuer.ref(defaultRef);

    private Cash.State createCashState(Amount<Currency> amount, AbstractParty owner) {
        return new Cash.State(issuedBy(amount, defaultIssuer), owner);
    }

    @Test
    public void mustIncludeSettleCommand() {
        Cash.State inputCash = createCashState(DOLLARS(5), bob.getParty());
        Cash.State outputCash = (Cash.State) inputCash.withNewOwner(alice.getParty()).getOwnableState();
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.input(Cash.PROGRAM_ID, inputCash);
                tx.output(Cash.PROGRAM_ID, outputCash);
                tx.command(bob.getPublicKey(), new Cash.Commands.Move());
                tx.fails();
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.input(Cash.PROGRAM_ID, inputCash);
                tx.output(Cash.PROGRAM_ID, outputCash);
                tx.command(bob.getPublicKey(), new Cash.Commands.Move());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new DummyCommand()); // Wrong type.
                tx.fails();
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.input(Cash.PROGRAM_ID, inputCash);
                tx.output(Cash.PROGRAM_ID, outputCash);
                tx.command(bob.getPublicKey(), new Cash.Commands.Move());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle()); // Correct Type.
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void mustHaveOnlyOneInputObligation() {
        Obligation duplicateObligation = new Obligation(DOLLARS(10), alice.getParty(), bob.getParty());
        Cash.State tenDollars = createCashState(DOLLARS(10), bob.getParty());
        Cash.State fiveDollars = createCashState(DOLLARS(5), bob.getParty());
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.failsWith("There must be one input obligation.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(OBLIGATION_CONTRACT_ID, duplicateObligation);
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.input(Cash.PROGRAM_ID, fiveDollars);
                tx.output(Cash.PROGRAM_ID, fiveDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), new Cash.Commands.Move());
                tx.failsWith("There must be one input obligation.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.input(Cash.PROGRAM_ID, tenDollars);
                tx.output(Cash.PROGRAM_ID, tenDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), new Cash.Commands.Move());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void mustBeCashOutputStatesPresent() {
        Cash.State cash = createCashState(DOLLARS(5), bob.getParty());
        CommandAndState cashPayment = cash.withNewOwner(alice.getParty());
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("There must be output cash.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, cash);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.output(Cash.PROGRAM_ID, cashPayment.getOwnableState());
                tx.command(bob.getPublicKey(), cashPayment.getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void mustBeCashOutputStatesWithRecipientAsOwner() {
        Cash.State cash = createCashState(DOLLARS(5), bob.getParty());
        CommandAndState invalidCashPayment = cash.withNewOwner(charlie.getParty());
        CommandAndState validCashPayment = cash.withNewOwner(alice.getParty());
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, cash);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.output(Cash.PROGRAM_ID, invalidCashPayment.getOwnableState());
                tx.command(bob.getPublicKey(), invalidCashPayment.getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("There must be output cash paid to the recipient.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, cash);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.output(Cash.PROGRAM_ID, validCashPayment.getOwnableState());
                tx.command(bob.getPublicKey(), validCashPayment.getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void cashSettlementAmountMustBeLessThanTheRemainingAmount() {
        Cash.State elevenDollars = createCashState(DOLLARS(11), bob.getParty());
        Cash.State tenDollars = createCashState(DOLLARS(10), bob.getParty());
        Cash.State fiveDollars = createCashState(DOLLARS(5), bob.getParty());
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, elevenDollars);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(11)));
                tx.output(Cash.PROGRAM_ID, elevenDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), elevenDollars.withNewOwner(alice.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("The amount settled cannot be more than the amount outstanding.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, fiveDollars);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.output(Cash.PROGRAM_ID, fiveDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), fiveDollars.withNewOwner(alice.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, tenDollars);
                tx.output(Cash.PROGRAM_ID, tenDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), tenDollars.withNewOwner(alice.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void cashSettlementMustBeInTheCorrectCurrency() {
        Cash.State tenDollars = createCashState(DOLLARS(10), bob.getParty());
        Cash.State tenPounds = createCashState(POUNDS(10), bob.getParty());
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, tenPounds);
                tx.output(Cash.PROGRAM_ID, tenPounds.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), tenPounds.withNewOwner(alice.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("Token mismatch: GBP vs USD");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, tenDollars);
                tx.output(Cash.PROGRAM_ID, tenDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), tenDollars.withNewOwner(alice.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void mustHaveOutputObligationIfNotFullySettling() {
        Cash.State tenDollars = createCashState(DOLLARS(10), bob.getParty());
        Cash.State fiveDollars = createCashState(DOLLARS(5), bob.getParty());
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, fiveDollars);
                tx.output(Cash.PROGRAM_ID, fiveDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), fiveDollars.withNewOwner(bob.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("There must be one output obligation.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, fiveDollars);
                tx.output(Cash.PROGRAM_ID, fiveDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.command(bob.getPublicKey(), fiveDollars.withNewOwner(bob.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(Cash.PROGRAM_ID, tenDollars);
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(10)));
                tx.output(Cash.PROGRAM_ID, tenDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), tenDollars.withNewOwner(bob.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("There must be no output obligation as it has been fully settled.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(Cash.PROGRAM_ID, tenDollars);
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(Cash.PROGRAM_ID, tenDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.command(bob.getPublicKey(), tenDollars.withNewOwner(bob.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void onlyPaidPropertyMayChange() {
        Cash.State fiveDollars = createCashState(DOLLARS(5), bob.getParty());
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, fiveDollars);
                tx.output(Cash.PROGRAM_ID, fiveDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.output(OBLIGATION_CONTRACT_ID, new Obligation(tenDollarObligation.getAmount(), tenDollarObligation.getLender(), alice.getParty(), DOLLARS(5)));
                tx.command(bob.getPublicKey(), fiveDollars.withNewOwner(bob.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("The borrower may not change when settling.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, fiveDollars);
                tx.output(Cash.PROGRAM_ID, fiveDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.output(OBLIGATION_CONTRACT_ID, new Obligation(DOLLARS(0), tenDollarObligation.getLender(), tenDollarObligation.getBorrower(), DOLLARS(5)));
                tx.command(bob.getPublicKey(), fiveDollars.withNewOwner(bob.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("The amount may not change when settling.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, fiveDollars);
                tx.output(Cash.PROGRAM_ID, fiveDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.output(OBLIGATION_CONTRACT_ID, new Obligation(DOLLARS(10), charlie.getParty(), tenDollarObligation.getBorrower(), DOLLARS(5)));
                tx.command(bob.getPublicKey(), fiveDollars.withNewOwner(bob.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("The lender may not change when settling.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.input(Cash.PROGRAM_ID, fiveDollars);
                tx.output(Cash.PROGRAM_ID, fiveDollars.withNewOwner(alice.getParty()).getOwnableState());
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.command(bob.getPublicKey(), fiveDollars.withNewOwner(bob.getParty()).getCommand());
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void mustBeSignedByAllParticipants() {
        Cash.State cash = createCashState(DOLLARS(5), bob.getParty());
        CommandAndState cashPayment = cash.withNewOwner(alice.getParty());
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(Cash.PROGRAM_ID, cash);
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(Cash.PROGRAM_ID, cashPayment.getOwnableState());
                tx.command(bob.getPublicKey(), cashPayment.getCommand());
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.command(ImmutableList.of(alice.getPublicKey(), charlie.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.failsWith("Both lender and borrower together only must sign obligation settle transaction.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(Cash.PROGRAM_ID, cash);
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(Cash.PROGRAM_ID, cashPayment.getOwnableState());
                tx.command(bob.getPublicKey(), cashPayment.getCommand());
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.command(bob.getPublicKey(), new ObligationContract.Commands.Settle());
                tx.failsWith("Both lender and borrower together only must sign obligation settle transaction.");
                return null;
            });
            ledger.transaction(tx -> {
                tx.input(Cash.PROGRAM_ID, cash);
                tx.input(OBLIGATION_CONTRACT_ID, tenDollarObligation);
                tx.output(Cash.PROGRAM_ID, cashPayment.getOwnableState());
                tx.command(bob.getPublicKey(), cashPayment.getCommand());
                tx.output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(DOLLARS(5)));
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ObligationContract.Commands.Settle());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }
}

File: obligation-cordapp/kotlin-source/src/test/kotlin/net/corda/examples/obligation/contract/ObligationContractSettleTests.kt
package net.corda.examples.obligation.contract

import net.corda.core.contracts.Amount
import net.corda.core.identity.AbstractParty
import net.corda.core.identity.CordaX500Name
import net.corda.core.utilities.OpaqueBytes
import net.corda.examples.obligation.Obligation
import net.corda.examples.obligation.ObligationContract
import net.corda.examples.obligation.ObligationContract.Companion.OBLIGATION_CONTRACT_ID
import net.corda.finance.DOLLARS
import net.corda.finance.POUNDS
import net.corda.finance.`issued by`
import net.corda.finance.contracts.asset.Cash
import net.corda.testing.core.TestIdentity
import net.corda.testing.node.ledger
import org.junit.Test
import java.util.*

class ObligationContractSettleTests : ObligationContractUnitTests() {
    private val issuer = TestIdentity(CordaX500Name("MegaBank", "", "US"))
    private val defaultRef = Byte.MAX_VALUE
    private val defaultIssuer = issuer.ref(defaultRef)

    private fun createCashState(amount: Amount<Currency>, owner: AbstractParty): Cash.State {
        return Cash.State(amount = amount `issued by` defaultIssuer, owner = owner)
    }

    @Test
    fun `must include settle command`() {
        val inputCash = createCashState(5.DOLLARS, bob.party)
        val outputCash = inputCash.withNewOwner(newOwner = alice.party).ownableState
        ledgerServices.ledger {
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                input(OBLIGATION_CONTRACT_ID, inputCash)
                output(OBLIGATION_CONTRACT_ID, outputCash)
                command(bob.publicKey, Cash.Commands.Move())
                this.fails()
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                input(OBLIGATION_CONTRACT_ID, inputCash)
                output(OBLIGATION_CONTRACT_ID, outputCash)
                command(bob.publicKey, Cash.Commands.Move())
                command(listOf(alice.publicKey, bob.publicKey), DummyCommand()) // Wrong type.
                this.fails()
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                input(OBLIGATION_CONTRACT_ID, inputCash)
                output(OBLIGATION_CONTRACT_ID, outputCash)
                command(bob.publicKey, Cash.Commands.Move())
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle()) // Correct Type.
                verifies()
            }
        }
    }

    @Test
    fun `must have only one input obligation`() {
        val duplicateObligation = Obligation(10.DOLLARS, alice.party, bob.party)
        val tenDollars = createCashState(10.DOLLARS, bob.party)
        val fiveDollars = createCashState(5.DOLLARS, bob.party)
        ledgerServices.ledger {
            transaction {
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                this `fails with` "There must be one input obligation."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, duplicateObligation)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                input(OBLIGATION_CONTRACT_ID, fiveDollars)
                output(OBLIGATION_CONTRACT_ID, fiveDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, Cash.Commands.Move())
                this `fails with` "There must be one input obligation."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                input(OBLIGATION_CONTRACT_ID, tenDollars)
                output(OBLIGATION_CONTRACT_ID, tenDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, Cash.Commands.Move())
                verifies()
            }
        }
    }

    @Test
    fun `must be cash output states present`() {
        val cash = createCashState(5.DOLLARS, bob.party)
        val cashPayment = cash.withNewOwner(newOwner = alice.party)
        ledgerServices.ledger {
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "There must be output cash."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, cash)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                output(OBLIGATION_CONTRACT_ID, cashPayment.ownableState)
                command(bob.publicKey, cashPayment.command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
        }
    }

    @Test
    fun `must be cash output states with receipient as owner`() {
        val cash = createCashState(5.DOLLARS, bob.party)
        val invalidCashPayment = cash.withNewOwner(newOwner = charlie.party)
        val validCashPayment = cash.withNewOwner(newOwner = alice.party)
        ledgerServices.ledger {
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, cash)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                output(OBLIGATION_CONTRACT_ID, invalidCashPayment.ownableState)
                command(bob.publicKey, invalidCashPayment.command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "There must be output cash paid to the recipient."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, cash)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                output(OBLIGATION_CONTRACT_ID, validCashPayment.ownableState)
                command(bob.publicKey, validCashPayment.command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
        }
    }

    @Test
    fun `cash settlement amount must be less than the remaining amount`() {
        val elevenDollars = createCashState(11.DOLLARS, bob.party)
        val tenDollars = createCashState(10.DOLLARS, bob.party)
        val fiveDollars = createCashState(5.DOLLARS, bob.party)
        ledgerServices.ledger {
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, elevenDollars)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(11.DOLLARS))
                output(OBLIGATION_CONTRACT_ID, elevenDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, elevenDollars.withNewOwner(newOwner = alice.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "The amount settled cannot be more than the amount outstanding."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, fiveDollars)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                output(OBLIGATION_CONTRACT_ID, fiveDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, fiveDollars.withNewOwner(newOwner = alice.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, tenDollars)
                output(OBLIGATION_CONTRACT_ID, tenDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, tenDollars.withNewOwner(newOwner = alice.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
        }
    }

    @Test
    fun `cash settlement must be in the correct currency`() {
        val tenDollars = createCashState(10.DOLLARS, bob.party)
        val tenPounds = createCashState(10.POUNDS, bob.party)
        ledgerServices.ledger {
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, tenPounds)
                output(OBLIGATION_CONTRACT_ID, tenPounds.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, tenPounds.withNewOwner(newOwner = alice.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "Token mismatch: GBP vs USD"
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, tenDollars)
                output(OBLIGATION_CONTRACT_ID, tenDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, tenDollars.withNewOwner(newOwner = alice.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
        }
    }

    @Test
    fun `must have output obligation if not fully settling`() {
        val tenDollars = createCashState(10.DOLLARS, bob.party)
        val fiveDollars = createCashState(5.DOLLARS, bob.party)
        ledgerServices.ledger {
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, fiveDollars)
                output(OBLIGATION_CONTRACT_ID, fiveDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, fiveDollars.withNewOwner(newOwner = bob.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "There must be one output obligation."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, fiveDollars)
                output(OBLIGATION_CONTRACT_ID, fiveDollars.withNewOwner(newOwner = alice.party).ownableState)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                command(bob.publicKey, fiveDollars.withNewOwner(newOwner = bob.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollars)
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(10.DOLLARS))
                output(OBLIGATION_CONTRACT_ID, tenDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, tenDollars.withNewOwner(newOwner = bob.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "There must be no output obligation as it has been fully settled."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollars)
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, tenDollars.withNewOwner(newOwner = alice.party).ownableState)
                command(bob.publicKey, tenDollars.withNewOwner(newOwner = bob.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
        }
    }

    @Test
    fun `only paid property may change`() {
        val fiveDollars = createCashState(5.DOLLARS, bob.party)
        ledgerServices.ledger {
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, fiveDollars)
                output(OBLIGATION_CONTRACT_ID, fiveDollars.withNewOwner(newOwner = alice.party).ownableState)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.copy(borrower = alice.party, paid = 5.DOLLARS))
                command(bob.publicKey, fiveDollars.withNewOwner(newOwner = bob.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "The borrower may not change when settling."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, fiveDollars)
                output(OBLIGATION_CONTRACT_ID, fiveDollars.withNewOwner(newOwner = alice.party).ownableState)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.copy(amount = 0.DOLLARS, paid = 5.DOLLARS))
                command(bob.publicKey, fiveDollars.withNewOwner(newOwner = bob.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "The amount may not change when settling."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, fiveDollars)
                output(OBLIGATION_CONTRACT_ID, fiveDollars.withNewOwner(newOwner = alice.party).ownableState)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.copy(lender = charlie.party, paid = 5.DOLLARS))
                command(bob.publicKey, fiveDollars.withNewOwner(newOwner = bob.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "The lender may not change when settling."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                input(OBLIGATION_CONTRACT_ID, fiveDollars)
                output(OBLIGATION_CONTRACT_ID, fiveDollars.withNewOwner(newOwner = alice.party).ownableState)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                command(bob.publicKey, fiveDollars.withNewOwner(newOwner = bob.party).command)
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
        }
    }

    @Test
    fun `must be signed by all participants`() {
        val cash = createCashState(5.DOLLARS, bob.party)
        val cashPayment = cash.withNewOwner(newOwner = alice.party)
        ledgerServices.ledger {
            transaction {
                input(OBLIGATION_CONTRACT_ID, cash)
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, cashPayment.ownableState)
                command(bob.publicKey, cashPayment.command)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                command(listOf(alice.publicKey, charlie.publicKey), ObligationContract.Commands.Settle())
                this `fails with` "Both lender and borrower together only must sign obligation settle transaction."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, cash)
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, cashPayment.ownableState)
                command(bob.publicKey, cashPayment.command)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                command(bob.publicKey, ObligationContract.Commands.Settle())
                this `fails with` "Both lender and borrower together only must sign obligation settle transaction."
            }
            transaction {
                input(OBLIGATION_CONTRACT_ID, cash)
                input(OBLIGATION_CONTRACT_ID, tenDollarObligation)
                output(OBLIGATION_CONTRACT_ID, cashPayment.ownableState)
                command(bob.publicKey, cashPayment.command)
                output(OBLIGATION_CONTRACT_ID, tenDollarObligation.pay(5.DOLLARS))
                command(listOf(alice.publicKey, bob.publicKey), ObligationContract.Commands.Settle())
                verifies()
            }
        }
    }
}